File: android/src/ti/modules/titanium/api/APIModule.java
package ti.modules.titanium.api;

import org.appcelerator.titanium.TiContext;
import org.appcelerator.titanium.TiDict;
import org.appcelerator.titanium.TiModule;
import org.appcelerator.titanium.util.Log;

public class APIModule extends TiModule
{
	private static final String LCAT = "TiAPI";

	private static TiDict constants;

	public static final int TRACE = 1;
	public static final int DEBUG = 2;
	public static final int INFO = 3;
	public static final int NOTICE = 4;
	public static final int WARN = 5;
	public static final int ERROR = 6;
	public static final int CRITICAL = 7;
	public static final int FATAL = 8;

	public APIModule(TiContext tiContext) {
		super(tiContext);
	}


	@Override
	public TiDict getConstants()
	{
		if (constants == null) {
			constants = new TiDict();

			constants.put("TRACE", TRACE);
			constants.put("DEBUG", DEBUG);
			constants.put("INFO", INFO);
			constants.put("NOTICE", NOTICE);
			constants.put("WARN", WARN);
			constants.put("ERROR", ERROR);
			constants.put("CRITICAL", CRITICAL);
			constants.put("FATAL", FATAL);
		}

		return constants;
	}

	public void debug(String msg) {
		Log.d(LCAT, msg);
	}

	public void info(String msg) {
		Log.i(LCAT, msg);
	}

	public void warn(String msg) {
		Log.w(LCAT, msg);
	}

	public void error(String msg) {
		Log.e(LCAT, msg);
	}

	public void trace(String msg) {
		Log.d(LCAT, msg);
	}

	public void notice(String msg) {
		Log.i(LCAT, msg);
	}

	public void critical(String msg) {
		Log.e(LCAT, msg);
	}

	public void fatal(String msg) {
		Log.e(LCAT, msg);
	}

	public void log(String level, String msg)
	{
		String ulevel = level.toUpperCase();
		int severity = INFO;

		if ("TRACE".equals(ulevel)) {
			severity = TRACE;
		} else if ("DEBUG".equals(ulevel)) {
			severity = DEBUG;
		} else if ("INFO".equals(ulevel)) {
			severity = INFO;
		} else if ("NOTICE".equals(ulevel)) {
			severity = NOTICE;
		} else if ("WARN".equals(ulevel)) {
			severity = WARN;
		} else if ("ERROR".equals(ulevel)) {
			severity = ERROR;
		} else if ("CRITICAL".equals(ulevel)) {
			severity = CRITICAL;
		} else if ("FATAL".equals(ulevel)) {
			severity = FATAL;
		} else {
			msg = "[" + level + "] " + msg;
		}

		internalLog(severity, msg);
	}

	public void internalLog(int severity, String msg)
	{
		if (severity == TRACE)
		{
			Log.v(LCAT,msg);
		}
		else if (severity < INFO)
		{
			Log.d(LCAT,msg);
		}
		else if (severity < WARN)
		{
			Log.i(LCAT,msg);
		}
		else if (severity == WARN)
		{
			Log.w(LCAT,msg);
		}
		else
		{
			Log.e(LCAT,msg);
		}
	}
}

File: iphone/TitaniumKit/TitaniumKit/Sources/API/APIModule.m
/**
 * Titanium SDK
 * Copyright TiDev, Inc. 04/07/2022-Present. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */
#import "APIModule.h"
#import "TiBase.h"
#import "TiExceptionHandler.h"

@implementation APIModule

- (NSString *)apiName
{
  return @"Ti.API";
}

- (bool)isNSBoolean:(id)object
{
  return [object isKindOfClass:[@YES class]];
}

- (void)logMessage:(id)args severity:(NSString *)severity
{
  if (args == nil) {
    args = @[ [NSNull null] ];
  } else if (!([args isKindOfClass:[NSArray class]])) {
    args = @[ args ];
  }
  // If the arg is an NSNumber wrapping a BOOL we should print the string equivalent for the boolean!
  NSMutableArray *newArray = [NSMutableArray array];
  [args enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
    if ([obj isKindOfClass:[JSValue class]]) {
      obj = ((JSValue *)obj).toObject;
    }
    if ([self isNSBoolean:obj]) {
      obj = [obj boolValue] ? @"true" : @"false";
    } else if (obj == nil) {
      obj = [NSNull null];
    }
    [newArray addObject:obj];
  }];
  NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
  [dateFormatter setDateFormat:@"yyyy-MM-dd HH:mm:ss"];
  // or @"yyyy-MM-dd hh:mm:ss a" if you prefer the time with AM/PM

  NSLog(@"[%@] %@ %@", [severity uppercaseString], [dateFormatter stringFromDate:[NSDate date]], [newArray componentsJoinedByString:@" "]);
}

- (id)transform:(id)arg
{
  if ([arg isKindOfClass:[NSDictionary class]]) {
    return [[[[TiScriptError alloc] initWithDictionary:arg] autorelease] description];
  }
  return arg;
}

- (void)debug:(id)args
{
  args = [JSContext currentArguments] ?: args;

  [self logMessage:args severity:@"debug"];
}

- (void)info:(id)args
{
  args = [JSContext currentArguments] ?: args;
  [self logMessage:args severity:@"info"];
}

- (void)warn:(id)args
{
  args = [JSContext currentArguments] ?: args;
  [self logMessage:args severity:@"warn"];
}

- (void)error:(NSArray *)args
{
  args = [JSContext currentArguments] ?: args;
  [self logMessage:args severity:@"error"];
}

- (void)trace:(id)args
{
  args = [JSContext currentArguments] ?: args;
  [self logMessage:args severity:@"trace"];
}

- (void)timestamp:(id)args
{
  id firstArg = args;
  if ([args isKindOfClass:[NSArray class]]) {
    firstArg = [args objectAtIndex:0];
  }
  NSLog(@"[TIMESTAMP] %f %@", [NSDate timeIntervalSinceReferenceDate], [self transform:firstArg]);
  fflush(stderr);
}

- (void)notice:(id)args
{
  args = [JSContext currentArguments] ?: args;
  [self logMessage:args severity:@"info"];
}

- (void)critical:(id)args
{
  args = [JSContext currentArguments] ?: args;
  [self logMessage:args severity:@"error"];
}

- (void)log:(id)level withMessage:(id)args
{
  if ([JSContext currentArguments].count > 0) {
    NSMutableArray *array = [NSMutableArray arrayWithArray:[JSContext currentArguments]];
    [array removeObjectAtIndex:0];
    args = array;
  }
  if (args == nil || ([args isKindOfClass:[NSArray class]] && [args count] == 0)) {
    [self logMessage:level severity:@"info"];
  } else {
    [self logMessage:args severity:level];
  }
}

- (void)reportUnhandledException:(NSArray *)args
{
  id lineNumber = [args objectAtIndex:0];
  id source = [args objectAtIndex:1];
  id message = [args objectAtIndex:2];

  NSLog(@"[ERROR] %@:%@ %@", source, lineNumber, message);
  fflush(stderr);
}

@end